clientlibrary/checkpoint/dynamodb-leasestealer.go

```diff
@@ -0,0 +1,190 @@
+//
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+package checkpoint
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbiface"
+
+	"github.com/vmware/vmware-go-kcl/clientlibrary/config"
+	par "github.com/vmware/vmware-go-kcl/clientlibrary/partition"
+	"github.com/vmware/vmware-go-kcl/logger"
+)
+
+// DynamoLeasestealer implements the Leasestealer interface using DynamoDB as a backend
+type DynamoLeasestealer struct {
+	log          logger.Logger
+	TableName    string
+	svc          dynamodbiface.DynamoDBAPI
+	kclConfig    *config.KinesisClientLibConfiguration
+	Retries      int
+	checkpointer Checkpointer
+}
+
+func NewDynamoLeasestealer(kclConfig *config.KinesisClientLibConfiguration, checkpointer Checkpointer) *DynamoLeasestealer {
+	leasestealer := &DynamoLeasestealer{
+		log:          kclConfig.Logger,
+		TableName:    kclConfig.TableName,
+		kclConfig:    kclConfig,
+		Retries:      NumMaxRetries,
+		checkpointer: checkpointer,
+	}
+
+	return leasestealer
+}
+
+// WithDynamoDB is used to provide DynamoDB service
+func (leasestealer *DynamoLeasestealer) WithDynamoDB(svc dynamodbiface.DynamoDBAPI) *DynamoLeasestealer {
+	leasestealer.svc = svc
+	return leasestealer
+}
+
+// Init initialises the DynamoDB Leasestealer
+func (leasestealer *DynamoLeasestealer) Init() error {
+	leasestealer.log.Infof("Creating DynamoDB session")
+	if leasestealer.svc == nil {
+		s, err := session.NewSession(&aws.Config{
+			Region:      aws.String(leasestealer.kclConfig.RegionName),
+			Endpoint:    aws.String(leasestealer.kclConfig.DynamoDBEndpoint),
+			Credentials: leasestealer.kclConfig.DynamoDBCredentials,
+			Retryer:     client.DefaultRetryer{NumMaxRetries: leasestealer.Retries},
+		})
+
+		if err != nil {
+			// no need to move forward
+			leasestealer.log.Fatalf("Failed in getting DynamoDB session for creating Worker: %+v", err)
+		}
+		leasestealer.svc = dynamodb.New(s)
+		return nil
+	}
+
+	return nil
+}
+
+// ListActiveWorkers returns a map of workers and their shards
+func (leasestealer *DynamoLeasestealer) ListActiveWorkers(shardStatus map[string]*par.ShardStatus) (map[string][]*par.ShardStatus, error) {
```

Comment:
https://github.com/patrobinson/gokini/blob/master/checkpointer.go can manage to put everything into one file and I think it's better to put into multiple files but under the same golang struct.

Reply:
What struct are you suggesting this live on? I could see it being on the Worker struct.

Reply:
Inside DynamoCheckpoint just like gokini.